File: core/src/main/java/org/mifos/mobilewallet/core/domain/usecase/UploadKYCDocs.java
package org.mifos.mobilewallet.core.domain.usecase;

import android.net.Uri;

import org.mifos.mobilewallet.core.base.UseCase;
import org.mifos.mobilewallet.core.data.fineract.entity.KYCDocsEnity;
import org.mifos.mobilewallet.core.data.fineract.repository.FineractRepository;

import javax.inject.Inject;

import okhttp3.ResponseBody;
import rx.Scheduler;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by ankur on 16/May/2018
 */

public class UploadKYCDocs extends UseCase<UploadKYCDocs.RequestValues,
        UploadKYCDocs.ResponseValue> {

    private final FineractRepository apiRepository;

    @Inject
    public UploadKYCDocs(FineractRepository apiRepository) {
        this.apiRepository = apiRepository;
    }

    @Override
    protected void executeUseCase(UploadKYCDocs.RequestValues requestValues) {
        KYCDocsEnity kycDocsEnity = new KYCDocsEnity();
        kycDocsEnity.setUri(requestValues.uri);
        apiRepository.uploadKYCDocs(requestValues.clientId, kycDocsEnity)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new Subscriber<ResponseBody>() {
                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable e) {
                        getUseCaseCallback().onError("Error upload Documents.");
                    }

                    @Override
                    public void onNext(ResponseBody responseBody) {
                        getUseCaseCallback().onSuccess(new ResponseValue());
                    }
                });
    }

    public static final class RequestValues implements UseCase.RequestValues {
        private final Uri uri;
        private final long clientId;

        public RequestValues(long clientId, Uri uri) {
            this.clientId = clientId;
            this.uri = uri;
        }
    }

    public static final class ResponseValue implements UseCase.ResponseValue {
        public ResponseValue() {
        }
    }
}

File: core/data/src/main/java/org/mifospay/core/data/domain/usecase/kyc/UploadKYCDocs.kt
/*
 * Copyright 2024 Mifos Initiative
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * See https://github.com/openMF/mobile-wallet/blob/master/LICENSE.md
 */
package org.mifospay.core.data.domain.usecase.kyc

import okhttp3.MultipartBody
import org.mifospay.core.data.base.UseCase
import org.mifospay.core.data.fineract.repository.FineractRepository
import org.mifospay.core.network.GenericResponse
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import javax.inject.Inject

class UploadKYCDocs @Inject constructor(
    private val apiRepository: FineractRepository,
) : UseCase<UploadKYCDocs.RequestValues, UploadKYCDocs.ResponseValue>() {

    class RequestValues(
        val entityType: String,
        val clientId: Long,
        val docname: String,
        val identityType: String,
        val file: MultipartBody.Part,
    ) : UseCase.RequestValues

    class ResponseValue : UseCase.ResponseValue

    override fun executeUseCase(requestValues: RequestValues) {
        apiRepository.uploadKYCDocs(
            requestValues.entityType,
            requestValues.clientId,
            requestValues.docname,
            requestValues.identityType,
            requestValues.file,
        )
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(
                object : Subscriber<GenericResponse>() {
                    override fun onCompleted() {}
                    override fun onError(e: Throwable) {
                        useCaseCallback.onError(e.toString())
                    }

                    override fun onNext(t: GenericResponse) {
                        useCaseCallback.onSuccess(ResponseValue())
                    }
                },
            )
    }
}
